Add osc resize

kargakis opened the pull request:
```
[vagrant@openshiftdev sample-app]$ osc describe dc/php
Name:		php
Created:	About an hour ago
Labels:		<none>
Latest Version:	2
Triggers:	Config, Image(php@latest, auto=true)
Strategy:	Recreate
Template:
	Selector:	deploymentconfig=php
	Replicas:	1
	Containers:
		NAME	IMAGE			ENV
		php	library/php:latest	
Latest Deployment:
	Name:		php-2
	Status:		Running
	Selector:	deployment=php-2,deploymentconfig=php
	Labels:		deployment=php-2,deploymentconfig=php
	Replicas:	0 current / 0 desired
	Pods Status:	0 Running / 0 Waiting / 0 Succeeded / 0 Failed
No events.

[vagrant@openshiftdev sample-app]$ osc describe rc/php-2
Name:		php-2
Image(s):	library/php:latest
Selector:	deployment=php-2,deploymentconfig=php
Labels:		deployment=php-2,deploymentconfig=php
Replicas:	0 current / 0 desired
Pods Status:	0 Running / 0 Waiting / 0 Succeeded / 0 Failed
No events.

[vagrant@openshiftdev sample-app]$ osc resize dc php --replicas=1
resized

[vagrant@openshiftdev sample-app]$ osc describe rc/php-2
Name:		php-2
Image(s):	library/php:latest
Selector:	deployment=php-2,deploymentconfig=php
Labels:		deployment=php-2,deploymentconfig=php
Replicas:	1 current / 1 desired
Pods Status:	1 Running / 0 Waiting / 0 Succeeded / 0 Failed
Events:
  FirstSeen				LastSeen			Count	From				SubobjectPath	Reason			Message
  Mon, 11 May 2015 10:20:24 +0000	Mon, 11 May 2015 10:20:24 +0000	1	{replication-controller }			successfulCreate	Created pod: php-2-zrbr4
```

cc: @smarterclayton @ironcladlou

mfojtik commented on pkg/deploy/resizer/resize.go:
```diff
@@ -0,0 +1,112 @@
+	}
+	if waitForReplicas != nil {
+		rc := &kapi.ReplicationController{ObjectMeta: kapi.ObjectMeta{Namespace: namespace, Name: rcName}}
+		if err := wait.Poll(waitForReplicas.Interval, waitForReplicas.Timeout,
```
nuke this `if`

mfojtik commented on pkg/deploy/resizer/resize.go:
```diff
@@ -0,0 +1,112 @@
+		return "", kubectl.ControllerResizeError{kubectl.ControllerResizeUpdateFailure, controller.ResourceVersion, err}
+	}
+	// TODO: do a better job of printing objects here.
+	return "resized", nil
```
can you turn this into a constant?

mfojtik commented:
LGTM, just a minor coding nits

mfojtik commented:
@kargakis one last thing... when you get a time after this is merged, can you please replace content that has this syntax:

```
$ osc update rc mongo --patch='{ "apiVersion": "v1beta1", "desiredState": { "replicas": 4 }}'
```

This example is from https://github.com/openshift/mongodb/tree/master/2.4/examples/replica#replicationcontroller-mongo but I do believe there are many more in our docs.

kargakis commented:
@mfojtik sure

kargakis commented:
I was passing the same resizerClient in both dc and rc resizing (rc is handled upstream so I was passing an invalid client there) but now it's fixed.
```
[vagrant@openshiftdev sample-app]$ osc resize rc php-2 --replicas=2
resized

[vagrant@openshiftdev sample-app]$ osc describe rc php-2
Name:		php-2
Image(s):	library/php:latest
Selector:	deployment=php-2,deploymentconfig=php
Labels:		deployment=php-2,deploymentconfig=php
Replicas:	2 current / 2 desired
Pods Status:	2 Running / 0 Waiting / 0 Succeeded / 0 Failed
Events:
  FirstSeen				LastSeen			Count	From				SubobjectPath	Reason			Message
  Mon, 11 May 2015 10:20:24 +0000	Mon, 11 May 2015 10:20:24 +0000	1	{replication-controller }			successfulCreate	Created pod: php-2-zrbr4
  Mon, 11 May 2015 11:42:15 +0000	Mon, 11 May 2015 11:42:15 +0000	1	{replication-controller }			successfulCreate	Created pod: php-2-izz24

[vagrant@openshiftdev sample-app]$ osc resize dc php --current-replicas=2 --replicas=3
resized

[vagrant@openshiftdev sample-app]$ osc describe rc php-2
Name:		php-2
Image(s):	library/php:latest
Selector:	deployment=php-2,deploymentconfig=php
Labels:		deployment=php-2,deploymentconfig=php
Replicas:	3 current / 3 desired
Pods Status:	3 Running / 0 Waiting / 0 Succeeded / 0 Failed
Events:
  FirstSeen				LastSeen			Count	From				SubobjectPath	Reason			Message
  Mon, 11 May 2015 10:20:24 +0000	Mon, 11 May 2015 10:20:24 +0000	1	{replication-controller }			successfulCreate	Created pod: php-2-zrbr4
  Mon, 11 May 2015 11:42:15 +0000	Mon, 11 May 2015 11:42:15 +0000	1	{replication-controller }			successfulCreate	Created pod: php-2-izz24
  Mon, 11 May 2015 12:06:46 +0000	Mon, 11 May 2015 12:06:46 +0000	1	{replication-controller }			successfulCreate	Created pod: php-2-o57eu
```

ironcladlou commented on pkg/deploy/resizer/resize.go:
```diff
@@ -0,0 +1,115 @@
+	if err != nil {
+		return nil, err
+	}
+	rcName = name + "-" + strconv.Itoa(dc.LatestVersion)
```
Use [LatestDeploymentNameForConfig](https://github.com/openshift/origin/blob/master/pkg/deploy/util/util.go#L20) instead,

kargakis commented:
I have updated the Kubernetes change to be on par with the PR[1] I have submitted upstream and addressed your comments @smarterclayton 

[1] https://github.com/GoogleCloudPlatform/kubernetes/pull/8057

smarterclayton commented:
LGTM [merge]

smarterclayton commented:
Test failed with

```
Error: Resizing the controller failed with: replicationControllers "ruby-hello-world-1" not found; Current resource version Unknown
!!! Error in hack/test-cmd.sh:431
  'osc resize dc ruby-hello-world --replicas=1' exited with status 1
Call stack:
  1: hack/test-cmd.sh:431 main(...)
```

kargakis commented:
```
[vagrant@openshiftdev origin]$ osc new-app https://github.com/openshift/ruby-hello-world -l name=hello; osc resize dc ruby-hello-world --replicas=1
imageStreams/ruby-hello-world
buildConfigs/ruby-hello-world
deploymentConfigs/ruby-hello-world
services/ruby-hello-world
A build was created - you can run `osc start-build ruby-hello-world` to start it.
Service "ruby-hello-world" created at 172.30.251.55 with port mappings 8080.
resized

[vagrant@openshiftdev origin]$ osc get rc
CONTROLLER           CONTAINER(S)       IMAGE(S)                          SELECTOR                                                                     REPLICAS
ruby-hello-world-1   ruby-hello-world   library/ruby-hello-world:latest   deployment=ruby-hello-world-1,deploymentconfig=ruby-hello-world,name=hello   1
```
:confused:

smarterclayton commented:
Let me know when this passes and I'll hit merge on it.

kargakis commented:
[test]

kargakis commented:
@smarterclayton test-cmd.sh passed on both ci environments, mind merging it?

smarterclayton commented:
[merge]

kargakis commented:
Rebased on top of current master and removed the upstream commit since it's already in after the latest Kube rebase. @mfojtik mind merging this?

mfojtik commented:
[merge]

smarterclayton commented:
[merge]
